map_gui/src/lib.rs

```diff
@@ -0,0 +1,342 @@
+                        drawn_all_areas = true;
+                    }
+                }
+                _ => {}
+            }
+
+            if self.current_selection == Some(obj.get_id()) {
+                g.draw_polygon(self.cs.selected, obj.get_outline(&self.map));
+            }
+        }
+    }
+
+    /// Assumes some defaults.
+    pub fn recalculate_current_selection(&mut self, ctx: &EventCtx) {
+        self.current_selection = self.calculate_current_selection(ctx, false, false);
+    }
+
+    pub fn mouseover_unzoomed_roads_and_intersections(&self, ctx: &EventCtx) -> Option<ID> {
+        self.calculate_current_selection(ctx, true, false)
+    }
+    pub fn mouseover_unzoomed_buildings(&self, ctx: &EventCtx) -> Option<ID> {
+        self.calculate_current_selection(ctx, false, true)
+    }
+    pub fn mouseover_unzoomed_everything(&self, ctx: &EventCtx) -> Option<ID> {
+        self.calculate_current_selection(ctx, true, true)
+    }
+
+    fn calculate_current_selection(
+        &self,
+        ctx: &EventCtx,
+        unzoomed_roads_and_intersections: bool,
+        unzoomed_buildings: bool,
+    ) -> Option<ID> {
+        // Unzoomed mode. Ignore when debugging areas.
+        if ctx.canvas.cam_zoom < self.opts.min_zoom_for_detail
+            && !(unzoomed_roads_and_intersections || unzoomed_buildings)
+        {
+            return None;
+        }
+
+        let pt = ctx.canvas.get_cursor_in_map_space()?;
+
+        let mut objects = self.draw_map.get_renderables_back_to_front(
+            Circle::new(pt, Distance::meters(3.0)).get_bounds(),
+            &self.map,
+        );
+        objects.reverse();
+
+        for obj in objects {
+            match obj.get_id() {
+                ID::Road(_) => {
+                    if !unzoomed_roads_and_intersections
+                        || ctx.canvas.cam_zoom >= self.opts.min_zoom_for_detail
+                    {
+                        continue;
+                    }
+                }
+                ID::Intersection(_) => {
+                    if ctx.canvas.cam_zoom < self.opts.min_zoom_for_detail
+                        && !unzoomed_roads_and_intersections
+                    {
+                        continue;
+                    }
+                }
+                ID::Building(_) => {
+                    if ctx.canvas.cam_zoom < self.opts.min_zoom_for_detail && !unzoomed_buildings {
+                        continue;
+                    }
+                }
+                _ => {
+                    if ctx.canvas.cam_zoom < self.opts.min_zoom_for_detail {
+                        continue;
+                    }
+                }
+            }
+            if obj.contains_pt(pt, &self.map) {
+                return Some(obj.get_id());
+            }
+        }
+        None
+    }
+}
+
+impl AppLike for SimpleApp {
+    #[inline]
+    fn map(&self) -> &Map {
+        &self.map
+    }
+    #[inline]
+    fn sim(&self) -> &Sim {
+        unreachable!()
```

Comment:
You eventually want to pull `Sim` out of map_gui and into game, right?

Reply:
Correct. There's a few funny dependencies, like `DrawIntersection` needing to know how much time is left in a traffic signal, and `ID` including some cases for agents. The former is easy enough to solve by the `sim_time` method, and one idea for the latter is to have separate `ID` enums here and in `game`, with methods for converting between the two.